File: 03.programmers/lv02/01.C++/LCMofN.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int GCD(int a, int b){
    if(a%b==0)return b;
    return GCD(b, a%b);
}
int LCM(int a, int b){return (a*b)/GCD(a,b);}
int solution(vector<int> arr) {
    int answer = arr[0];   
    for(int i=1; i<arr.size(); i++)
        answer = LCM(answer,arr[i]);

    return answer;
}
int main(){
    cout<<solution({4,6,20})<<endl;
    return 0;
}

// int solution(vector<int> arr) {
//     int answer =0;
//     sort(arr.begin(), arr.end());
//     int chk = 0;
//     while(chk!=arr.size()){
//         answer += arr[0];
//         chk=0;
//         for(int i=0; i<arr.size(); i++)
//             if(answer % arr[i] == 0)chk++;
//     }

//     return answer;
// }

File: ~2023/03.programmers/lv02/02.Python/LCM_of_N.py
def solution(arr):
    answer = arr[0]
    GCD = lambda a, b : b if a%b==0 else GCD(b,a%b)
    LCM = lambda a, b : int(a*b/GCD(a,b))
    for i in range(1,len(arr)):
        answer = LCM(answer,arr[i])
    return answer

input_arr = [2,6,8,14]
print(solution(input_arr))

# def GCD(a,b): return b if a%b==0 else GCD(b,a%b)
# def LCM(a,b): return int((a*b)/GCD(a,b))
# def solution(arr):
#     answer = arr[0]
#     for i in range(1,len(arr)):
#         answer = LCM(answer,arr[i])
#     return answer